scipy/signal/filter_design.py

```diff
@@ -329,21 +343,47 @@ def freqz(b, a=1, worN=None, whole=False, plot=None):
 
     """
     b, a = map(atleast_1d, (b, a))
-    if whole:
-        lastpoint = 2 * pi
-    else:
-        lastpoint = pi
-    if worN is None:
-        N = 512
-        w = numpy.linspace(0, lastpoint, N, endpoint=False)
-    elif isinstance(worN, int):
-        N = worN
-        w = numpy.linspace(0, lastpoint, N, endpoint=False)
+
+    worN = 512 if worN is None else worN  # historically we allowed None
+
+    h = None
+    if isinstance(worN, int):
+        lastpoint = 2 * pi if whole else pi
+        w = numpy.linspace(0, lastpoint, worN, endpoint=False)
+        min_size = max(a.shape[axis], b.shape[axis])
+        if worN >= min_size and fftpack.next_fast_len(worN) == worN:
+            # if worN is fast, 2 * worN will be fast, too,
+            # so no need to check that one
+            n_fft = worN if whole else worN * 2
+            if np.isrealobj(b) and np.isrealobj(a):
+                h = np.fft.rfft(b, n=n_fft, axis=axis)[:worN]
+                if a.size == 1:
+                    h /= a
+                else:
+                    h /= np.fft.rfft(a, n=n_fft, axis=axis)[:worN]
+                if whole:
+                    # exclude DC and maybe Nyquist
+                    stop = h.shape[axis] - 1 if n_fft % 2 == 1 else -2
+                    h_flip = [slice(None)] * h.ndim
+                    h_flip[axis] = slice(stop, 0, -1)
+                    h = np.concatenate((h, h[tuple(h_flip)].conj()), axis=axis)
+            else:  # we can use complex FFT
+                h = fftpack.fft(b, n=n_fft, axis=axis)[:worN]
+                if a.size == 1:
+                    h /= a
+                else:
+                    h /= fftpack.fft(a, n=n_fft, axis=axis)[:worN]
     else:
-        w = worN
-    w = atleast_1d(w)
-    zm1 = exp(-1j * w)
-    h = polyval(b[::-1], zm1) / polyval(a[::-1], zm1)
+        w = atleast_1d(worN)
+    del worN
+
+    if h is None:  # still need to compute using freqs w
+        zm1 = exp(-1j * w)
+        b_flip = [slice(None)] * b.ndim
+        b_flip[axis] = slice(None, None, -1)
+        a_flip = [slice(None)] * a.ndim
+        a_flip[axis] = slice(None, None, -1)
+        h = polyval(b[tuple(b_flip)], zm1) / polyval(a[tuple(a_flip)], zm1)
```

Comment:
This block after `if h is None:` doesn't quite work yet:
```
In [1]: from scipy.signal import freqz, butter

In [2]: b = np.random.rand(2, 7, 3)

In [3]: a = np.random.rand(2, 7, 3)

In [4]: w, h = freqz(b, a, worN=5, axis=1)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-4-821898c3ee79> in <module>()
----> 1 w, h = freqz(b, a, worN=5, axis=1)

~/miniconda3scipy/lib/python3.5/site-packages/scipy-1.0.0.dev0+f3fb55e-py3.5-macosx-10.6-x86_64.egg/scipy/signal/filter_design.py in freqz(***failed resolving arguments***)
    388         a_flip = [slice(None)] * a.ndim
    389         a_flip[axis] = slice(None, None, -1)
--> 390         h = polyval(b[tuple(b_flip)], zm1) / polyval(a[tuple(a_flip)], zm1)
    391     if plot is not None:
    392         plot(w, h)

~/miniconda3scipy/lib/python3.5/site-packages/numpy/lib/polynomial.py in polyval(p, x)
    678         y = NX.zeros_like(x)
    679     for i in range(len(p)):
--> 680         y = y * x + p[i]
    681     return y
    682 

ValueError: operands could not be broadcast together with shapes (5,) (7,3) 
```
It looks like it could be replaced with

```
    zm1 = exp(-1j * w)
    bb = b.swapaxes(axis, 0)
    aa = a.swapaxes(axis, 0)
    h = (npp_polyval(zm1, bb) / npp_polyval(zm1, aa))
```